### ISSUE TITLE
Add possibility to define default sort for Panache entities

### ISSUE THREAD

**Issue by marceloverdijk** [quarkusio/quarkus#6439]
As discussed briefly https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/quarkus-dev/ktb7DF9a08g/yPk_aDfzDAAJ it would be useful to define a default sort for Panache entities.

This default sort should then be used implicitly by e.g. `list*()`, `find*()` when no explicit sorting/ordering is provided.

**Comment by FroMage** [comment]
This should be trivial with an annotation on the class. Pity that https://docs.oracle.com/javaee/7/api/javax/persistence/OrderBy.html is not allowed on a class.

This can be useful for mongo too, right @loicmathieu ?

**Comment by FroMage** [comment]
Given that neither JPA's `@OrderBy` and Hibernate's `@Sort` are allowed on classes, do @emmanuelbernard or @Sanne have ideas for an annotation name we can use that would not clash with the existing ones?

**Comment by loicmathieu** [comment]
@FroMage yeah, sure if we define a `@DefaultSort` annotation (or whatever we named it) this should be easy to use it for both Hibernate and MongoDB with Panache.

Idealy don't use `@DefaultOrderBy` as it referes to the SQL ORDER BY clause and Panache try to use higher level abstraction ...

**Comment by FroMage** [comment]
Huh, actually we could pervert the `@OrderBy` annotation to place it on the field we want to sort by, rather than on the entity class. But that'd only work for single field sort…

**Comment by loicmathieu** [comment]
Please use `@SortBy` and it can works on mutiple field if you add an optional `ordinal` property on the annotation to know in which order you need to sort. Each sort field will be annotated ...

**Comment by emmanuelbernard** [comment]
I don't understand the use case. You want every single possible find/list operations to always be using a uniquely defined sort order for a given entity type? I am having a hard time to see this as the real use case. 

I generally find *static* (aka annotation) definitions for such settings the incorrect approach.

What will we do for queries? If they don't override sort, will it be used? How will you detect that?

**Comment by marceloverdijk** [comment]
The use case is to define a default natural order for an entity. And yes I would expect that all query methods would use this default sort if not overridden.

The alternative is that I basically in all query where no "specific" order is needed, I have to pass the default sort my self. Which is cumbersome, but could also be forgotten.

Other ORM tools like Grails support such a feature as well: see e.g. http://docs.grails.org/3.1.1/ref/Database%20Mapping/order.html

**Comment by emmanuelbernard** [comment]
I find it a bit surprising that for some entity you always want an order. But it maye make sense for focused apps.

Ideally that would be a Hibernate feature @Sanne ? But we could make it a panache one temporarily.

I think `@DefaultOrderBy` makes sense as it mimics https://javaee.github.io/javaee-spec/javadocs/javax/persistence/OrderBy.html which is about attribute level ordering (@loicmathieu I think you got confused on that one).

```
@DefaultOrderBy("lastname, firstname")
public class User extends PanacheEntity {
}
```

But that would apply on all queries unless order is specify which means some tricky work for Panache? 

@marceloverdijk do you want to try and write a PoC so we see the feasibility and whether the semantics are what we want?

**Comment by FroMage** [comment]
> I find it a bit surprising that for some entity you always want an order. But it maye make sense for focused apps.

Yeah, not sure how useful this would be, perhaps it's premature to add it without waiting for more user requests?

> But that would apply on all queries unless order is specify which means some tricky work for Panache?

My vote would go to applying it on all queries without order indeed. Not hard to implement.

**Comment by marceloverdijk** [comment]
> My vote would go to applying it on all queries without order indeed.

Yes, that's exactly what I would expect.

**Comment by Sanne** [comment]
> My vote would go to applying it on all queries without order indeed.

Same, which in my mind makes this more a Panache feature than an Hibernate feature; incidentally, it would also be far harder in Hibernate so if people feel it belongs in Hibernate I'd indeed wait for many more votes.

**Comment by FroMage** [comment]
Yeah, I think panache is a good low-risk incubation layer. We don't need to add stuff in Hibernate until we try them there.

**Comment by emmanuelbernard** [comment]
My reserve was that "Panache" queries will be covered but not non panache ones. So if I fall back to EM APIs, I need to apply my filter. Maybe that's acceptable.

**Comment by fiorenzino** [comment]
It's really common to define a default order for single db table view (using rest api) and offering to the web user a way to change the field to re-order (also with asc or desc).
A visual use case:
https://primefaces.org/primeng/showcase/#/table/sort
When you load the data for the first time, the order could be a developer choice.

My +1 for:
@DefaultOrderBy("lastname asc, firstname desc") at class level, always overridable by Sort in all queries.

**Comment by tanadeau** [comment]
I think this would be a useful feature for an application that I'm working on. Most of our entities have a natural, default sort like a title or two title-like fields.
